backpack/extensions/module_extension.py

```diff
@@ -12,98 +24,156 @@ class ModuleExtension:
+            and is_loss(module) is False
+            and TORCH_VERSION_HIGHER_THAN_1_9_0
+        ):
+            raise AssertionError(
+                "BackPACK extension expects a backpropagation quantity but it is None. "
+                f"Module: {module}, Extension: {extension}."
+            )
 
         for param in self.__params:
             if self.__param_exists_and_requires_grad(module, param):
                 extFunc = getattr(self, param)
-                extValue = extFunc(ext, module, g_inp, g_out, bpQuantities)
-                self.__save(extValue, ext, module, param)
+                extValue = extFunc(extension, module, g_inp, g_out, bp_quantity)
+                self.__save_value_on_parameter(extValue, extension, module, param)
+
+        if extension.expects_backpropagation_quantities():
+            bp_quantity = self.backpropagate(
+                extension, module, g_inp, g_out, bp_quantity
+            )
+            self.__save_backprop_quantity(
+                extension,
+                module.input0,
+                bp_quantity,
+            )
 
-        bpQuantities = self.backpropagate(ext, module, g_inp, g_out, bpQuantities)
+    @staticmethod
+    def __get_backproped_quantity(
+        extension: BackpropExtension, reference_tensor: Tensor
+    ) -> Tensor or None:
+        """Fetch backpropagated quantities attached to the module output.
 
-        self.__backprop_quantities(ext, inp, out, bpQuantities)
+        Args:
+            extension: current BackPACK extension
+            reference_tensor: the output Tensor of the current module
 
-    @staticmethod
-    def __backproped_quantities(ext, out):
-        """Fetch backpropagated quantities attached to the module output."""
-        return getattr(out, ext.savefield, None)
+        Returns:
+            the backpropagation quantity
+        """
+        return extension.saved_quantities.retrieve_quantity(reference_tensor.data_ptr())
 
     @staticmethod
-    def __backprop_quantities(ext, inp, out, bpQuantities):
-        """Propagate back additional information by attaching it to the module input."""
-
-        setattr(inp, ext.savefield, bpQuantities)
+    def __save_backprop_quantity(
+        extension: BackpropExtension,
+        reference_tensor: Tensor,
+        bpQuantities: Any,
+    ) -> None:
+        """Propagate back additional information by attaching it to the module input.
+
+        Args:
+            extension: current BackPACK extension
+            reference_tensor: reference tensor on which to save
+            bpQuantities: backpropagation quantities that should be saved
+        """
+        extension.saved_quantities.save_quantity(
+            reference_tensor.data_ptr(), bpQuantities
+        )
 
-        is_a_leaf = out.grad_fn is None
-        retain_grad_is_on = getattr(out, "retains_grad", False)
-        inp_is_out = id(inp) == id(out)
-        should_retain_grad = is_a_leaf or retain_grad_is_on or inp_is_out
```

Comment:
Needs to be introduces. Discuss in person.

Reply:
Reintroduce, then this will be resolved.

Reply:
Reintroduced. But I had to comment out `inp_is_out` since it breaks the current implementation.

Reply:
Do you have an idea why it breaks?

Reply:
Yes.
Modules like ReLU have same input and output.
Old: inp_is_out=True -> should_retain_grad=True -> gets overwritten it in the next module
New (with inp_is_out): inp_is_out=True -> should_retain_grad=True -> raises Error in nextModule: Quantitiy already exists, to the code it looks like a ResNet but ResNets are not supported yet
New (without inp_is_out): should_retain_grad=False -> quantity gets deleted

I'm not sure how to resolve the conflict between retain_grad=True and ResNets, because they both require quantities to have the same save_vector(old) or data_ptr (new).

Reply:
Can you post the output of this snippet for `torch>=1.9.0`?
```python
  from torch import rand
  from torch.nn import ReLU
  from backpack import extend

  a = rand(10, 10)
  layer = extend(ReLU())
  b = layer(a)

  print(id(layer.input0) == id(layer.output))
```
(I get `False` on `1.7.1`)

Reply:
`False` on `1.9.0`

Reply:
Seems like my description above is completely wrong and the issue is somewhere else.

Reply:
Waiting for more information. Maybe requires in-person discussion

Reply:
The above statement holds, but for `Flatten` instead of `ReLU`.
```Python
from torch import rand
from torch.nn import Flatten
from backpack import extend

a = rand(10, 10)
layer = extend(Flatten())
b = layer(a)

print(id(layer.input0) == id(layer.output))
```
This is `True` only if `Flatten` does nothing. Change `a=rand(10, 10, 10)` and the ids are different.

Reply:
I think I understand the problem:

If `Flatten` does not perform an operation, it should simply retain the backpropagated quantity and skip its own `backpropagate` method. 

Your current code does not retain the backpropagated quantity and calls `Flatten`'s `backpropagate` function.

I tried out the first approach in [this commit](https://github.com/f-dangel/backpack/commit/668e3aa343075da17b163b7e09b521312deec768) and was able to get rid of a lot of additional boilerplate in `Flatten`'s second-order extensions. Can you take a look at the diff and try to spot any flaws? If it looks good, you can merge it into your branch.

Reply:
Thought a bit more about it. Approach 2 also implies that `backpropagate` of activation functions with `inplace=True`, like `Relu(inplace=True)` will not be called because `id(input0) == id(output)`. But in this case an exception should be raised that such inplace operations are not supported by BackPACK. Hence, approach 2 could lead to situations where BackPACK computes wrong results but the code passes.

Having thought about those edge cases, I think we can stay with your current (1.) solution and leave `inp_is_out` commented out.

Marking as resolved. Reopen in case of follow-up questions.